### PR TITLE
Update recompute_bin_data for file backed pixel data

### LINKED ISSUE #281: Update recompute_bin_data for file-backed PixelData
The `recompute_bin_data` algorithm does not work if all pixel data is not in memory. In the worst case, the C++ algorithm segfaults and Matlab crashes. This needs to be updated before any unary operations can be implemented for file-backed data.

**Outputs:**
- `recompute_bin_data` returns the same result if the pixel data is file-backed or all held in memory
- some tests to this effect

### PR DESCRIPTION
**Pull request by hsautessella**
This PR updates the `recompute_bin_data` algorithm for use on file-backed pixel data. 

The implementation has been moved into the `PixelData` class. The Matlab and C++ versions have been updated.

**Note:**
There's an algorithm [`average_bin_data`](https://github.com/hsautessella/Horace/blob/d4bb3f858e4ff56d6e453b7b3d2eeff9a029631a/horace_core/sqw/@sqw/private/average_bin_data.m) that is similar to `recompute_bin_data` that will most likely need updating as well. I haven't looked too deeply into it, but we shouldn't have two implementations of a similar algorithm. If possible, this function should be updated to call out to `PixelData.compute_bin_data`.

Fixes #281

### CODE REVIEW

File: horace_core/sqw/PixelData/@PixelData/PixelData.m [hunk 1]
```diff
@@ -435,7 +440,7 @@ function delete(obj)
         % and clear the current cache
         %  This function does nothing if pixels are not file-backed.
         %
-        if obj.is_file_backed_()
+        if obj.is_file_backed_() && obj.page_number_ ~= 1
```
[the remark this answers is not in the document]
Comment (hsautessella) [reply horace_core/sqw/PixelData/@PixelData/PixelData.m:443]: Is there anything to write or persist? If you're on page one and:
- there's no data in the cache - no changes have been made and there's nothing to remove. 
- there's unedited data in the cache - doing nothing leaves the data as is.
- there's edited data in the cache - doing nothing leaves the data as is, and calling `advance` will still save any changes

File: _test/test_sqw/test_PixelData_operations.m [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    end
+
+    function test_compute_bin_data_correct_output_in_memory_mex_1_thread(obj)
+        obj.config.use_mex = true;
+        obj.config.threads = 1;
+
+        pix = PixelData(obj.test_sqw_file_path);
+        [s, e] = pix.compute_bin_data(obj.ref_npix_data);
+
+        assertEqual(s, obj.ref_s_data, '', 4.75e-4);
```
[the remark this answers is not in the document]
Comment (hsautessella) [reply _test/test_sqw/test_PixelData_operations.m:44]: The reference signal and variance data is different. The final number in the `assert` is a tolerance

File: horace_core/sqw/PixelData/@PixelData/PixelData.m (hunk 1 above)
[the remark this answers is not in the document]
Comment (hsautessella) [reply horace_core/sqw/PixelData/@PixelData/PixelData.m:443]: The reason I put this condition in, is that `move_to_first_page` will clear the cache and only reload if data is requested. If we're already on the first page, then we're just clearing the cache and then re-loading the data from the file

[the remark this answers is not in the document]
Comment (nickbattam-tessella) [reply horace_core/sqw/PixelData/@PixelData/PixelData.m:443]: Ok, that makes sense.

The "empty" `zeros(9,0)` below should be `zeros(PIXEL_BLOCK_COLS_ , 0)`

**Comment by hsautessella** [comment]
`average_bin_data` takes data as an argument and as such assumes data is in memory. Therefore, the algorithm does not need updating to use paged data at this time

**Pull request merged**